vm/ignite/src/runtime.rs

```diff
@@ -85,57 +119,125 @@ pub fn execute(rt: &mut Runtime, instr: ByteCode) -> Result<bool> {
         ByteCode::BINOP(op) => micro_code::binop(rt, op)?,
         ByteCode::JOF(pc) => micro_code::jof(rt, pc)?,
         ByteCode::GOTO(pc) => micro_code::goto(rt, pc)?,
-        ByteCode::RESET(t) => micro_code::reset(rt, t)?,
+        ByteCode::RESET(ft) => micro_code::reset(rt, ft)?,
         ByteCode::ENTERSCOPE(syms) => micro_code::enter_scope(rt, syms)?,
         ByteCode::EXITSCOPE => micro_code::exit_scope(rt)?,
         ByteCode::CALL(arity) => micro_code::call(rt, arity)?,
+        ByteCode::SPAWN => micro_code::spawn(rt)?,
+        ByteCode::JOIN(tid) => micro_code::join(rt, tid)?,
+        ByteCode::YIELD => micro_code::yield_(rt)?,
     }
-    Ok(false)
+    Ok(())
 }
 
-/// Extend the current environment with new symbols and values.
-///
-/// # Arguments
-///
-/// * `rt` - The runtime to extend the environment of.
-///
-/// * `syms` - The symbols to add to the environment.
-///
-/// * `vals` - The values to add to the environment.
-///
-/// # Errors
-///
-/// If the symbols and values are not the same length.
-pub fn extend_environment<S, V>(rt: &mut Runtime, syms: Vec<S>, vals: Vec<V>) -> Result<()>
-where
-    S: Into<Symbol>,
-    V: Into<Value>,
-{
-    if syms.len() != vals.len() {
-        return Err(VmError::IllegalArgument(
-            "symbols and values must be the same length".to_string(),
-        )
-        .into());
+impl Runtime {
+    /// Check if the time quantum has expired.
+    /// The time quantum is the maximum amount of time a thread can run before it is preempted.
+    pub fn time_quantum_expired(&self) -> bool {
+        self.time.elapsed() >= self.time_quantum
+    }
+
+    /// Check if the current thread should yield.
+    /// This is set by the `YIELD` instruction.
+    pub fn should_yield_current_thread(&self) -> bool {
+        self.current_thread.state == ThreadState::Yielded
+    }
+
+    /// Yield the current thread. Set the state of the current thread to `Ready` and push it onto the ready queue.
+    /// Pop the next thread from the ready queue and set it as the current thread.
+    /// The timer is reset to the current time.
+    pub fn yield_current_thread(mut self) -> Self {
+        let mut current_thread = self.current_thread;
+        current_thread.state = ThreadState::Ready; // Reset the state
+        self.ready_queue.push_back(current_thread);
+
+        let next_ready_thread = self
+            .ready_queue
+            .pop_front()
+            .expect("No threads in ready queue");
```

Comment:
If all threads get suspended that might be a nice way to do a simple deadlock detector 🤔 (for later / after the course)